gql_websocket_link/lib/src/link.dart

```diff
@@ -1,66 +1,208 @@
 import "dart:async";
+import "dart:convert";
 
 import "package:gql_exec/gql_exec.dart";
 import "package:gql_link/gql_link.dart";
-import "package:gql_websocket_link/src/get_from_ast.dart";
-import "package:gql_websocket_link/src/socket_client.dart";
 import "package:gql_websocket_link/src/messages.dart";
-import "package:gql/language.dart";
-import "package:gql/ast.dart";
+import "package:rxdart/rxdart.dart";
+import "package:uuid_enhanced/uuid.dart";
+import "package:web_socket_channel/io.dart";
 
 /// A Universal WebSocket [Link] implementation to support the WebSocket-GraphQL transport.
 /// It supports subscriptions, query and mutation operations as well.
 ///
 /// NOTE: the actual socket connection will only get established after a [Request] is handled by this [WSLink].
-/// If you'd like to connect to the socket server instantly, call the [connectOrReconnect] method after creating this [WSLink] instance.
 class WSLink extends Link {
-  final String url;
-  SocketClient _client;
-  final SocketClientConfig config;
+  String _uri;
+  IOWebSocketChannel _channel;
 
+  /// Serializer used to serialize request
+  final RequestSerializer serializer;
+
+  /// Response parser
   ResponseParser parser;
 
-  WSLink(
-    this.url, {
-    this.config = const SocketClientConfig(),
+  /// Payload to be sent with the connection_init request
+  /// Must be able to `json.encode(initialPayload)`.
+  final dynamic initialPayload;
+
+  // Possible states of the connection.
+  static const int connecting = 0;
+  static const int open = 1;
+  static const int closing = 2;
+  static const int closed = 3;
+  final BehaviorSubject<int> _connectionStateController =
+      BehaviorSubject<int>();
+
+  Stream<GraphQLSocketMessage> _messageStream;
+  StreamSubscription<GraphQLSocketMessage> _messageSubscription;
+
+  /// Initialize the WSLink with a [uri].
+  /// You can customize the headers & protocols by passing [channel],
+  /// if [channel] is passed, [uri] must be null.
+  /// You can also pass custom [RequestSerializer serializer] & [ResponseParser parser].
+  /// Also [initialPayload] to be passed with the first request to the GraphQL server.
+  WSLink({
+    String uri,
```

Comment:
Just because `HttpLink` has `uri` as a positional argument, I'd prefer to align this to the other link and make `uri` positional.